apiserver: refactor graceful termination logic

tkashem opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
- refactor graceful termination logic so we can write unit tests to assert on the expected behavior.
- add unit test(s) that assert on the expected behavior

It paves the way for https://github.com/kubernetes/kubernetes/pull/101257

#### Which issue(s) this PR fixes:

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

tkashem commented:
/assign @sttts @p0lyn0mial

tkashem commented:
ran this test using `stress`, so far looks like there is no flake. 

```
$ go test -race k8s.io/apiserver/pkg/server -c
$ stress ./server.test -test.run=TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled
...
19m50s: 7014 runs so far, 0 failures
19m55s: 7042 runs so far, 0 failures
20m0s: 7074 runs so far, 0 failures
20m5s: 7102 runs so far, 0 failures
20m10s: 7130 runs so far, 0 failures
```

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
+	}
+
+	serverShutdownCh, listenerStoppedCh, err := s.NonBlockingRun(events.WithName(AfterShutdownDelayDuration).Signaled())
 	if err != nil {
```
if `NonBlockingRun` returns an `error` we won't:
- stop the AuditBacked
- stop the `goroutine 1`
- stop the `goroutine 2`

tkashem replied:
we are aborting, the primary goroutine that is executing `Run` will abort and all other go routines will be aborted with it. I don't think we need to be graceful when `Run` returns an error, thoughts?

tkashem replied:
okay, I moved `s.AuditBackend.Run` to its original site, now it is in keeping with the current implementation

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 	if err != nil {
 		return err
 	}
+	klog.V(1).Info("[graceful-termination] RunPreShutdownHooks has completed")
```
should we convert to structured logging?

tkashem replied:
`[graceful-termination]` is just a prefix so we can search all graceful termination related log entries with this keyword. Is there a way to add a prefix in structured logging?

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
+		defer close(s.readinessStopCh)
 
-		time.Sleep(s.ShutdownDelayDuration)
+		<-events.WithName(ShutdownInitiated).Signaled()
```
mixing `the new abstraction` and `regular chan` makes it a bit hard to read

tkashem replied:
yes, i agree. I simplified a bit, let me know if it reads better. I guess this is a tradeoff, we need the abstraction for a flake free test, on the other hand, we lose nice idiomatic inline channel and go func.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 // or the secure port cannot be listened on initially.
 func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
-	delayedStopCh := make(chan struct{})
+	events := s.TerminationEvents
```
`events` implies the usage of the regular events, rename to `signals`?

tkashem replied:
i renamed to `TerminationSignal` and brought back the old variable names.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
+	// Start the audit backend before any request comes in. This means we must call Backend.Run
+	// before http server start serving. Otherwise the Backend.ProcessEvents call might block.
+	if s.AuditBackend != nil {
+		if err := s.AuditBackend.Run(events.WithName(InFlightRequestsDrained).Signaled()); err != nil {
```
I think that `InFlightRequestsDrained` doesn't guarantee that all in-flight requests have been processed

tkashem replied:
This is in keeping with the current behavior - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L419-L420

I think this is correct - `WithWaitGroup` precedes `WithAudit` - so existing requests in flight have already been added to the audit backend. so `auditBackend.Run` will stop as soon `WithWaitGroup` has finished executing all existing requests in flight, thoughts?

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
-		close(s.readinessStopCh)
+		// This gives the load balancer a window defined by ShutdownDelayDuration to
+		// detect that /readyz is red and stop sending traffic to this server.
+		defer close(s.readinessStopCh)
```
we are not interested in knowing that `close(readinessStopCh)` was closed?

tkashem replied:
I removed `readinessStopCh` from `GenericAPIServer`. we no longer need it, we can use `ShutdownInitiated` signal instead.
```
	// as soon as shutdown is initiated, readiness should start failing
	readinessStopCh := s.TerminationSignals.ShutdownInitiated.Signaled()
	err := s.addReadyzShutdownCheck(readinessStopCh)
```

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
-	delayedStopCh := make(chan struct{})
+	events := s.TerminationEvents
+
+	signalFn := s.overrideTerminationEventSignalFunc
```
so, looking at the previous code https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go it seems like knowing when `delayedStopCh` and `drainedCh` were closed would allow to write unit tests and cover the entire `Run` method.

I'm just thinking if we could restore the previous version of the code and inject a channel factory instead.
The factory would allow for creating a new chan and wrapping already existing ones. 

Then we would change 
the `Run` method to accept the factory `Run(factoryCh chanFactory)`.
`make(chan struct{})` to `factoryChan.new("name")`
`close(...)` to `chanWrapperInstance.close()`
reading from a chan `<-` to calling the `wait` method



```
type chanFactory interface {
 new(name) chanWrapper
 wrap(chan) chanWrapper
}


type chanWrapper interface {
 wait() // blocking 
 close()
 rawChan()
}
```

After that we would get:
```

func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}, factoryCh chanFactory) error {
	delayedStopCh := factoryCh.new("delayedStopCh")
         stopCh = factoryCh.wrap(stopCh)

	go func() {
		deferdelayedStopCh.close()

		stopCh.wait()

		// As soon as shutdown is initiated, /readyz should start returning failure.
		// This gives the load balancer a window defined by ShutdownDelayDuration to detect that /readyz is red
		// and stop sending traffic to this server.
		close(s.readinessStopCh)

		time.Sleep(s.ShutdownDelayDuration)
	}()

	// close socket after delayed stopCh
	stoppedCh, err := s.NonBlockingRun(delayedStopCh.rawChan())
	if err != nil {
		return err
	}

         stoppedCh = factoryCh.wrap(stoppedCh)

	drainedCh := factoryCh.new("drainedCh")
	go func() {
		defer drainedCh.close()

		// wait for the delayed stopCh before closing the handler chain (it rejects everything after Wait has been called).
		delayedStopCh.wait()

		// Wait for all requests to finish, which are bounded by the RequestTimeout variable.
		s.HandlerChainWaitGroup.Wait()
	}()

	stopCh.wait()

	// run shutdown hooks directly. This includes deregistering from the kubernetes endpoint in case of kube-apiserver.
	err = s.RunPreShutdownHooks()
	if err != nil {
		return err
	}

	// Wait for all requests in flight to drain, bounded by the RequestTimeout variable.
	drainedCh.wait()
	// wait for stoppedCh that is closed when the graceful termination (server.Shutdown) is finished.
	stoppedCh.wait()

	return nil
}
```

p0lyn0mial replied:
not sure how hard would it be to create these abstractions.

tkashem replied:
A couple of notes:
- i have a followup PR where i have a filter that will need the termination event channel, so we can't create these channels inside the `Run` function, see here https://github.com/kubernetes/kubernetes/pull/101257/files#diff-207875acc7ee03f987aa626f53742c02323ed8f434fd40dfd929dffb23d3f1deR789.
```
	handler = genericfilters.WithHSTS(handler, c.HSTSDirectives)
	if c.KeepListeningDuringGracefulTermination {
		handler = genericfilters.WithNotAcceptingNewRequest(handler, c.TerminationEvents.WithName(NotAcceptingNewRequest).Wait())
	}
	handler = genericfilters.WithHTTPLogging(handler)
``

-

tkashem replied:
Also, 
```
type chanWrapper interface {
 wait() // blocking 
 close()
 rawChan()
}
```

is similar to what we have in this PR
```
type TerminationSignal interface {
	Name() string
	Signal()
	Signaled() <-chan struct{}
}
```

- `wait` is similar to `Signaled` (except `Signaled` returns a `<- chan struct{}` as opposed to `wait` which blocks ) `Signaled` seems more idiomatic and more composable, 
- with `Signaled` we don't need `rawChan`
- `close` = `Signal`

If it makes it easier to read i can go with
```
type TerminationSignal interface {
	Name() string
	close()
	done() <-chan struct{}
}
```
Thoughts?

fedebongio commented:
/triage accepted

tkashem commented:
the test looks good, no flake seen in about `45m` run
```
$ go test -race k8s.io/apiserver/pkg/server -c
$ stress ./server.test -test.run=TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled
47m25s: 15330 runs so far, 0 failures
47m30s: 15358 runs so far, 0 failures
47m35s: 15386 runs so far, 0 failures
47m40s: 15410 runs so far, 0 failures
...

tkashem commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -219,6 +219,10 @@ type Config struct {
 
+	// TerminationSignals provides access to the various shutdown signals
+	// that happen during the graceful termination of the apiserver.
+	TerminationSignals *TerminationSignals
```
why is this a pointer? Is it optional?

tkashem replied:
it is never optional, i removed the pointer.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -213,6 +210,10 @@ type GenericAPIServer struct {
+
+	// TerminationSignals provides access to the various termination
+	// signals that happen during the shutdown period of the apiserver.
+	TerminationSignals *TerminationSignals
```
is this ever nil?

sttts replied:
does this have to be public ?

tkashem replied:
it should not be public, i made it private

sttts commented on staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go:
```diff
@@ -0,0 +1,144 @@
+
+	// Signaled returns a channel that is closed when the underlying termination
+	// event has been signaled. Successive calls to Signaled return the same value.
+	Signaled() <-chan struct{}
```
I like this type.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -390,20 +404,22 @@ func (s preparedGenericAPIServer) NonBlockingRun(stopCh <-chan struct{}) (<-chan
 		var err error
-		stoppedCh, err = s.SecureServingInfo.Serve(s.Handler, s.ShutdownTimeout, internalStopCh)
+		klog.V(1).Infof("[graceful-termination] ShutdownTimeout=%s", s.ShutdownTimeout)
+		stoppedCh, listenerStoppedCh, err = s.SecureServingInfo.Serve(s.Handler, s.ShutdownTimeout, internalStopCh)
```
all the changes around listenerStoppedCh might deserve its own commit. This changes the wiring.

tkashem replied:
i put this change in a new commit

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -330,58 +334,68 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
+	go func() {
+		<-listenerStoppedCh
+		httpServerStoppedListeningCh.Signal()
+	}()
```
move this below the block where listenerStoppedCh is returned.

tkashem replied:
done

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -330,58 +334,68 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
+	httpServerStoppedListeningCh := s.TerminationSignals.HTTPServerStoppedListening
+	go func() {
+		<-listenerStoppedCh
+		httpServerStoppedListeningCh.Signal()
```
do we want to log when the socket was closed ?

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go:
```diff
@@ -0,0 +1,144 @@
+func (e *namedChannelWrapper) Signal() {
+	select {
+	case <-e.ch:
+		// already closed, don't close again.
```
I would remove it and let in `panic` just like normal channels

tkashem replied:
if `Signal` is called more than once then the test would fail, the test actually verifies that the events are signaled exactly once and in the right order:
```
	terminationSignalOrderExpected := []string{
		string("ShutdownInitiated"),
		string("AfterShutdownDelayDuration"),
		string("HTTPServerStoppedListening"),
		string("InFlightRequestsDrained"),
	}
```

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+func (e *namedChannelWrapper) Name() string {
+	return string(e.name)
```
returning `e.name` is not enough?

tkashem replied:
fixed

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go:
```diff
@@ -0,0 +1,144 @@
+
+func newNamedChannelWrapper(name string) TerminationSignal {
+	return &namedChannelWrapper{
+		name: name,
```
isn't this used only in the tests? if yes then I'm wondering if it could be removed.

tkashem replied:
yes, it can be removed, it's just easier for the the test to get the name instead of using a type assertion.

p0lyn0mial replied:
I would remove it and add it to the embedded type in the test.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:
```diff
@@ -0,0 +1,399 @@
+
+	// record the termination events in the order they are signaled
+	var eventOrderLock sync.Mutex
+	eventOrderGot := make([]string, 0)
```
nit rename to `signalOrderGot` ?

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:
```diff
@@ -0,0 +1,399 @@
+	}
+
+	// step 2: signal termination event: initiate a shutdown
+	close(stopCh)
```
+1

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:
```diff
@@ -0,0 +1,399 @@
+	}
+
+	// wrap the termination signals of the GenericAPIServer so the test can inject its own callback
+	wrapTerminationSignals(t, s.TerminationSignals, func(before bool, e TerminationSignal) {
```
great, so now the test simpy catches events from the `Run` method

tkashem committed:
apiserver: refactor graceful termination logic

- refactor graceful termination logic so we can write unit tests
  to assert on the expected behavior.

tkashem committed:
apiserver: NonBlockingRun should return a listener stopped channel

NonBlockingRun should also return a channel that gets closed when the
underlying http Server has stopped listening (during the graceful
shutdown period)

tkashem committed:
add unit test to verify graceful termination behavior

tkashem commented:
/retest

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/secure_serving.go:
```diff
@@ -192,6 +192,67 @@ func (s *SecureServingInfo) Serve(handler http.Handler, shutdownTimeout time.Dur
+// It returns a listenerStoppedCh that is closed when the underlying http Server has stopped listening.
+// TODO: do a follow up PR to remove this function, change 'Serve' to return listenerStoppedCh
+//  and update all components that call 'Serve'
+func (s *SecureServingInfo) ServeWithListenerStopped(handler http.Handler, shutdownTimeout time.Duration, stopCh <-chan struct{}) (<-chan struct{}, <-chan struct{}, error) {
```
we are going to remove the `Serve` method in a follow-up PR.

tkashem replied:
yes, i have added a TODO for that as well

p0lyn0mial commented:
/lgtm

sttts commented:
/approve
